### ISSUE TITLE
UserAsync methods execute synchronously

### ISSUE THREAD

**Issue by Ahmed-Abdelhameed** [linvi/tweetinvi#670]
Hello,

I wanted to use the `User.GetUserFromScreenName()` method asynchronously, so I used the `UserAsync` class instead but I found out it executes synchronously too. Here's an example:

```c#
// This will execute synchronously.
var targetUser = await UserAsync.GetUserFromScreenName("username");
// This will execute asynchronously
var followers = await targetUser.GetFollowersAsync(1000);
```

Is this a bug, or am I doing something wrong?

Thanks!

**Comment by linvi** [comment]
Hello, I am not sure I understand your question.

You are using the `await` keyword. So the `UserAsync.GetUserFromScreenName` is being executed asynchronously but is awaited by the `await` keyword.

The `targetUser.GetFollowersAsync(1000)` will behave the same way.

Could you please explain a bit more clearly what you think is not going correctly here?

Thanks,
Linvi

**Comment by Ahmed-Abdelhameed** [comment]
Hello, thanks for replying.

Yes, that's what I'm expecting, but it doesn't execute asynchronously, instead, it executes exactly the same as `User.GetUserFromScreenName`, meaning it blocks the current thread (i.e., it executes synchronously).

**Comment by linvi** [comment]
Yes because you use the `await` keyword. Remove the await and it will not block anymore.

**Comment by Ahmed-Abdelhameed** [comment]
The `await` keyword does **Not** block the current thread! It simply schedules the rest of the code *in the method* to be executed after the asynchronous operation finishes. Any method that is fully asynchronous (e.g., `Task.Delay(xx)`, `HttpClient.GetAsync()`, etc.) should **never** block the current thread (e.g., the UI thread). Don't you agree with that?

**Comment by linvi** [comment]
Yes I agree with your statement about `async`. I have various developers opening issues (from absolute beginner to senior) so I was not sure if you meant that the Thread was blocking or if the execution was blocking.

Now that we are on the same page, I will investigate the problem.

Could you let me know which type of application you are developing as the `SynchronizationContext` is different based on the type of application you are using.

In addition to that could you confirm that you are using version 2.1.

Thanks

**Comment by Ahmed-Abdelhameed** [comment]
Hello, thanks for clarifying!

I'm currently using the library in a WinForms application, hence the *current thread* is the *UI thread*. And yes, I'm using version 2.1.0.

One more thing, the methods `GetUsersFromIds` and `GetUsersFromScreenNames` don't exist in the `UserAsync` class (they only exist in the `User` class). Is there no asynchronous implementation for them?

Thank you!

**Comment by linvi** [comment]
Tweetinvi is behind the scene relying on Thread Static. This design is obviously creating lots of challenges in regards to `async` implementation.

As a result, the `async` migration of the library is planned for a version 3.0 of the library which will not happen anytime soon as I have way too much to work on.

In the meantime I therefore implemented a helper static class `Sync`. You can then transform any operation into an `async` operation.

Here is an example : 

``` c#
var authenticatedUser = await Sync.ExecuteTaskAsync(() => { return User.GetAuthenticatedUser(); });
```

Normally any `async` operation provided as part of the `Async` static classes should use this behind the scene.

Anyway, I will look at this issue and give you additional information.

Thanks,
Linvi

**Comment by Ahmed-Abdelhameed** [comment]
I didn't check the source code, but AFAICT, `Sync.ExecuteTaskAsync` doesn't seem to do what you say.
The following will still get executed synchronously:

```C#
var users = await Sync.ExecuteTaskAsync(() => { return User.GetUsersFromIds(userIds); });
```

**Comment by linvi** [comment]
I can reproduce the problem with Winforms events.
I will look more into the problem, thanks for the precision and for reporting it.

**Comment by linvi** [comment]
This is now fixed and will be part of the version 2.2 of Tweetinvi.

Thanks for this good spot :)
